bindingcollectionadapter/src/main/java/me/tatarka/bindingcollectionadapter2/itembindings/OnItemBindClass.java

```diff
@@ -44,6 +74,26 @@ public int itemTypeCount() {
 
     @Override
     public void onItemBind(ItemBinding itemBinding, int position, T item) {
+        itemBinding.clearExtra();
+        if (itemExtraBindingMap != null && !itemExtraBindingMap.isEmpty()) {
+            for (int i = 0; i < itemExtraBindingMap.size(); i++) {
+                Class<? extends T> key = itemExtraBindingMap.keyAt(i);
+                if (key.isInstance(item)) {
```

Comment:
Sorry for my copy paste but I've done similar your code:

    /**
     * Maps the given class to the given variableId and layout. This is an exact match, no
     * inheritance it taken into account.
     */
    public OnItemBindClass<T> map(@NonNull Class<? extends T> itemClass, int variableId, @LayoutRes int layoutRes) {
        itemBindingMap.put(itemClass, new int[]{variableId, layoutRes});
        return this;
    }

    @Override
    public void onItemBind(ItemBinding itemBinding, int position, T item) {
        for (int i = 0; i < itemBindingMap.size(); i++) {
            Class<? extends T> key = itemBindingMap.keyAt(i);
            if (key.isInstance(item)) {
                int[] values = itemBindingMap.valueAt(i);
                itemBinding.set(values[0], values[1]);
                return;
            }
        }
        throw new IllegalArgumentException("Missing class for item " + item);
    }

    @Test
    public void selectsBasedOnSubclass() ...


So in your code call of method map() order dependent. Example:

    private static class A {
    }

    private static class B extends A {
    }

    private static class C {
    }

This is represent wrong behaviour because B mapping will never used:

        OnItemBindClass<Object> onItemBind = new OnItemBindClass<>()
                .map(A.class, 0, 1)
                .map(B.class, 4, 5)
                .map(C.class, 2, 3);

This is right usage:

        OnItemBindClass<Object> onItemBind = new OnItemBindClass<>()
                .map(B.class, 4, 5)
                .map(A.class, 0, 1)
                .map(C.class, 2, 3);